diag_manager/fms_doubly_linked_list.F90

```diff
@@ -0,0 +1,275 @@
+    procedure :: next => literator_next
+    procedure :: get => literator_data
+  end type literator
+
+  type, public :: linked_list
+    !! Note we are overriding the default constructor with an
+    ! interface of the same name
+    type(node), pointer :: head
+    type(node), pointer :: tail
+    integer :: theSize
+  contains
+    procedure :: push_front => push_at_front
+    procedure :: push_back => push_at_back
+    procedure :: pop_front => pop_at_front
+    procedure :: pop_back => pop_at_back
+    procedure :: insert => insert_data
+    procedure :: remove => remove_node
+    procedure :: get_literator => get_forward_literator
+    procedure :: size => get_size
+    procedure :: is_empty => is_size_zero
+    procedure :: clear => clear_all
+    final :: destructor
+  end type linked_list
+
+  interface node
+    module procedure :: node_constructor
+  end interface node
+
+  interface linked_list
+    module procedure :: linked_list_constructor
+  end interface linked_list
+
+  interface literator
+    module procedure :: literator_constructor
+  end interface literator
+
+contains
+
+  !> Insert data d in a new node to be placed in front of the
+  ! target node t_nd. Return an iterator that starts with the
+  ! newly inserted node.
+  function  insert_data( this, t_nd,  d ) result(liter)
+    class(linked_list), intent(in out) :: this
+    class(node), pointer, intent (in)  :: t_nd
+    class(*), target, intent(in)       :: d
+    class(literator), allocatable      :: liter
+    !!
+    class(node), pointer :: nd
+    allocate(nd)
+    nd%data => d
+    nd%prev => t_nd%prev
+    nd%next => t_nd
+    t_nd%prev%next => nd
+    t_nd%prev => nd
+    this%theSize = this%theSize + 1
+    liter = literator(nd, this%tail)
+  end function insert_data
+
+  !> Remove Node nd from the linked tree. Return the iterator
+  ! that begins with the nex node after nd, and ends with the
+  ! list end node. Return the list iterator if the node cannot
+  ! be removed.
+  function  remove_node( this, nd ) result( litr)
+    class(linked_list), intent(in out) :: this
+    type(node), pointer, intent(in out) :: nd
+    class(literator), ALLOCATABLE  :: litr
+    !Dont even try to remove the head and tail nodes!
+    if (.not. ( associated (this%head , nd ) .or. &
+      associated (this%tail , nd ) )) THEN
+      litr = literator( nd%next, this%tail )
+      nd%prev%next => nd%next
+      nd%next%prev => nd%prev
+      deallocate(nd)
+      this%theSize =  this%theSize - 1
+    else
+      litr = this%get_literator()
+    endif
+  end function remove_node
+
+
+  !>  Remove the data at the head of the list. Return an
+  !  Iterator starting with the next (new front) node, and
+  ! Return the list iterator if the list is already empty
+  function pop_at_front (this ) result( liter )
+    class(linked_list), intent(in out) :: this
+    class(literator), allocatable :: liter
+    !!
+    class(node), pointer :: nd
```

Comment:
Variables need to be documented in doxygen style

Reply:
Done!